src/core/qgspolymorphicrelation.h

```diff
@@ -0,0 +1,288 @@
+/***************************************************************************
+    qgspolymorphicrelation.h
+     --------------------------------------
+    Date                 : December 2020
+    Copyright            : (C) 2020 Ivan Ivanov
+    Email                : ivan at opengis dot ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSPOLYMORPHICRELATION_H
+#define QGSPOLYMORPHICRELATION_H
+
+#include <QList>
+#include <QDomNode>
+#include <QPair>
+
+#include "qgis_core.h"
+#include "qgsfields.h"
+#include "qgsreadwritecontext.h"
+#include "qgsrelationcontext.h"
+#include "qgsrelation.h"
+
+#include "qgis_sip.h"
+
+class QgsFeatureIterator;
+class QgsFeature;
+class QgsFeatureRequest;
+class QgsAttributes;
+class QgsVectorLayer;
+class QgsPolymorphicRelationPrivate;
+class QgsExpressionContext;
+
+/**
+ * A polymorphic relation consists of the same properties like an ordinary relation except for the referenced layer which is calculated based on one or several fields of the referencing layer.
```

Comment:
I think in other places it's referred to "normal" relations. I'd stick to one single terminology.